net processing: Move block inventory state to net_processing

jnewbery opened the pull request:
This continues the work of moving application layer data into net_processing, by moving all block inventory state into the new Peer object added in #19607.

For motivation, see #19398.

jnewbery committed:
[net processing] Rename nStartingHeight to m_starting_height

Not done as a scripted diff to avoid misnaming the local variable in
ProcessMessage().

jnewbery committed:
scripted-diff: rename vBlockHashesToAnnounce and vInventoryBlockToSend

-BEGIN VERIFY SCRIPT-
sed -i 's/vBlockHashesToAnnounce/m_blocks_for_headers_relay/g' src/net_processing.*
sed -i 's/vInventoryBlockToSend/m_blocks_for_inv_relay/g' src/net_processing.*
-END VERIFY SCRIPT-

jnewbery committed:
[net processing] Move nStartingHeight to Peer

jnewbery committed:
[net processing] Move block inventory data to Peer

jnewbery committed:
[net processing] Move hashContinue to net processing

Also rename to m_continuation_block to better communicate meaning.

jnewbery committed:
[net processing] Improve documentation for Peer destruction/locking

Suggested here:

- https://github.com/bitcoin/bitcoin/pull/19607#discussion_r467071878
- https://github.com/bitcoin/bitcoin/pull/19829#discussion_r546116786

ariard commented on src/net.h:
```diff
@@ -928,9 +922,6 @@ class CNode
     // m_tx_relay == nullptr if we're not relaying transactions with this peer
     std::unique_ptr<TxRelay> m_tx_relay;
 
-    // Used for headers announcements - unfiltered blocks to relay
```
What's the distinction you loose by dropping `unfiltered blocks to relay` ?

I guess this is a reference to sort done at headers announcement, L4218 src/net_processing.cpp:
https://github.com/bitcoin/bitcoin/blob/1cf73fb8eba3b89a30d5e943deaec5052a8b21b7/src/net_processing.cpp#L4218

Before to announce headers, we check both with regards to peer's context ("Does this header connect to known peer's tip ?") and our context ("Does this header is part of best chain ?"). It wasn't a clear comment, so yes either drop it or extend ?

jnewbery replied:
I think drop it.

ariard commented on src/net_processing.cpp:
```diff
@@ -488,6 +488,14 @@ struct Peer {
+     * There is no final sorting before sending, as they are always sent
+     * immediately and in the order requested. */
+    std::vector<uint256> vInventoryBlockToSend GUARDED_BY(m_block_inv_mutex);
+    /** List of headers to relay */
```
`vBlockHashesToAnnounce` is a list of block hashes not headers ? With regards to renaming follow-up commit, I would lean towards the most verbose option `m_blockids_for_inv_relay/m_blockids_for_headers_relay` to denotate that these vectors contain an identifier even if the endgoal is to relay the identifiee (either block or headers).

jnewbery replied:
Ok, I'll update the comment to "List of block ids to announce headers for".

> I would lean towards the most verbose option m_blockids_for_inv_relay/m_blockids_for_headers_relay

Sure. I'll update.

ariard reviewed (commented):
Just calls for better comments.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -488,6 +488,9 @@ struct Peer {
     bool m_should_discourage GUARDED_BY(m_misbehavior_mutex){false};
 
+    /** Starting height of this peer */
+    std::atomic<int> nStartingHeight{-1};
```
The comment doesn't need to document _everything_ that the variable is used for. That's what the code does.

jnewbery commented:
Thanks for the review @ariard . I've addressed your comments.

jnewbery commented:
Rebased

jonatack commented:
Concept ACK

hebasto commented:
Concept ACK.

hebasto commented on src/net_processing.cpp:
```diff
@@ -488,6 +488,14 @@ struct Peer {
     bool m_should_discourage GUARDED_BY(m_misbehavior_mutex){false};
 
+    /** Protects block inventory data members */
+    RecursiveMutex m_block_inv_mutex;
```
e04fb9a2df9f4d77c587bd25c608ce806c97ea45
Could using of `RecursiveMutex` type be avoided?
(as there was `Mutex cs_inventory;`)

jnewbery replied:
Good catch! `cs_inventory` was a RecursiveMutex when I opened this PR, and I missed the change when I rebased. Fixed.

hebasto replied:
#19347 is yours, btw :)

jnewbery replied:
:flushed:

hebasto commented on src/net_processing.cpp:
```diff
@@ -505,6 +505,24 @@ using PeerRef = std::shared_ptr<Peer>;
+template<typename Callable>
+void ForEachPeer(Callable&& func)
+{
+    std::vector<PeerRef> peer_copies;
+    {
+        LOCK(g_peer_mutex);
+        peer_copies.reserve(g_peer_map.size());
+        for (auto peer : g_peer_map) {
+            peer_copies.push_back(peer.second);
+        }
+    }
+
+    for (auto&& peer : peer_copies) {
+        func(peer);
```
575c5f55cc44348f0ebc7740c90013b96bdb8c27
It is good for concurrency to limit the lock scope. But is it safe to call `func` on potentially stale `peer_copies`?

jnewbery replied:
Yes. `peer_copies` holds shared_ptr references, so the objects pointed to can't be destructed while it's in scope.

hebasto replied:
I mean not language safety rather client behavior safety: if another thread removes a `peer` from the `g_peer_map`, can we rely on result of `func(peer)`?

jnewbery replied:
Yes, because peer still exists in that case.

jnewbery commented:
Thanks for the review @hebasto . I've addressed both of your comments.

hebasto reviewed (approved):
ACK 06841521613f1e7312f3f51597b43a20c193a363, tested on Linux Mint 20 (x86_64).

jnewbery commented:
rebased

jnewbery commented:
rebased

jnewbery commented:
rebased

jonatack reviewed (commented):
First pass through this follow-up to #19607. A few comments below; feel free to ignore the nits. My main concern is that several commits aren't hygienic as mentioned below and that we're missing a test that exhibits this without manually looking at `getpeerinfo` or the GUI peer window.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1542,6 +1582,7 @@ static void RelayAddress(const CAddress& addr, bool fReachable, const CConnman&
 
 void static ProcessGetBlockData(CNode& pfrom, const CChainParams& chainparams, const CInv& inv, CConnman& connman)
 {
+    PeerRef peer = GetPeerRef(pfrom.GetId());
```
Correct. Future commits will use `peer` throughout the function, and there's no downside to holding on to the object for a bit longer.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -913,6 +948,10 @@ void PeerManager::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
+        // a reference to the object. Be careful not to do any processing here that
+        // assumes Peer won't be changed before it's destructed.
         PeerRef peer = GetPeerRef(nodeid);
         assert(peer != nullptr);
```
Done

jnewbery commented on src/net_processing.cpp:
```diff
@@ -2334,6 +2375,7 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
         return;
     }
 
+    PeerRef peer = GetPeerRef(pfrom.GetId());
```
Yes, potentially we could just pass a `Peer&` to many of the private functions (just as `CNodeState*` is passed to some of the functions now).

jnewbery commented on src/net.h:
```diff
@@ -940,8 +940,6 @@ class CNode
 public:
-    uint256 hashContinue;
-    std::atomic<int> nStartingHeight{-1};
 
```
done

jnewbery commented on src/net_processing.cpp:
```diff
@@ -502,6 +519,24 @@ using PeerRef = std::shared_ptr<Peer>;
 
+/** Get a reference to each Peer, then call a function on each of them */
+template<typename Callable>
+void ForEachPeer(Callable&& func)
```
I don't have a strong opinion. I've done it this way here to minimize the difference in logic, but if people want to change the internal implementation later, I'm fine with that.

jnewbery commented on src/qt/rpcconsole.cpp:
```diff
@@ -1148,6 +1147,8 @@ void RPCConsole::updateNodeDetail(const CNodeCombinedStats *stats)
         else
             ui->peerCommonHeight->setText(tr("Unknown"));
+
+        ui->peerHeight->setText(QString::number(stats->nodeStateStats.m_starting_height));
```
Oops. I think this was a bad rebase. Should be fixed now.

jnewbery commented on src/rpc/net.cpp:
```diff
@@ -202,14 +202,14 @@ static RPCHelpMan getpeerinfo()
             }
             obj.pushKV("synced_headers", statestats.nSyncHeight);
             obj.pushKV("synced_blocks", statestats.nCommonHeight);
+            obj.pushKV("startingheight", statestats.m_starting_height);
```
As above, should now be fixed.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -488,6 +488,23 @@ struct Peer {
+    std::vector<uint256> m_blocks_for_headers_relay GUARDED_BY(m_block_inv_mutex);
+
+    /** Starting height of this peer */
+    std::atomic<int> m_starting_height{-1};
```
> m_starting_height presumably still needs to be std::atomic_int?

I don't understand. It's a `std::atomic<int>` before and after this PR.

> If you retouch, perhaps s/height/block height/ for the Doxygen doc.

I've updated the Doxygen comment to be clearer.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -4196,7 +4239,7 @@ bool PeerManager::SendMessages(CNode* pto)
                 // then send all headers past that one.  If we come across any
                 // headers that aren't on ::ChainActive(), give up.
-                for (const uint256 &hash : pto->vBlockHashesToAnnounce) {
+                for (const uint256 &hash : peer->m_blocks_for_headers_relay) {
```
done

jnewbery commented on src/net_processing.cpp:
```diff
@@ -488,6 +488,23 @@ struct Peer {
+
+    /** Starting height of this peer */
+    std::atomic<int> m_starting_height{-1};
+    /** The final block hash in a INV message. When the peer requests this
```
fixed

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1399,11 +1439,11 @@ void PeerManager::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInde
-            LOCK(pnode->cs_inventory);
-            if (nNewHeight > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : 0)) {
+        ForEachPeer([nNewHeight, &vHashes](const PeerRef& peer) {
+            LOCK(peer->m_block_inv_mutex);
```
Oops. Bad rebase. Should be fixed now.

jnewbery commented:
Thanks for the very thorough review, @jonatack. I've addressed all of your comments.

jonatack commented on src/net_processing.cpp:
```diff
@@ -971,6 +1011,7 @@ bool GetNodeStateStats(NodeId nodeid, CNodeStateStats &stats) {
     PeerRef peer = GetPeerRef(nodeid);
     if (peer == nullptr) return false;
     stats.m_misbehavior_score = WITH_LOCK(peer->m_misbehavior_mutex, return peer->m_misbehavior_score);
+    stats.m_starting_height = peer->m_starting_height;
```
0a9839058c
```bash
net_processing.cpp: In function ‘bool GetNodeStateStats(NodeId, CNodeStateStats&)’:
net_processing.cpp:982:11: error: ‘struct CNodeStateStats’ has no member named ‘m_starting_height’; did you mean ‘nStartingHeight’?
  982 |     stats.m_starting_height = peer->m_starting_height;
      |           ^~~~~~~~~~~~~~~~~
      |           nStartingHeight
```

jnewbery replied:
fixed

jonatack reviewed (commented):
The new changes (`git diff 8d3a068 1a9e191`) look good. One rebase fixup in 0a9839058c.

jnewbery commented:
Thanks @jonatack . I fixed the bad rebase.

jonatack commented:
commits f9136f4bf, 7199d81968, 8ff3e9319c11 and 40ea4a42d9
```diff
net_processing.cpp: In function ‘bool GetNodeStateStats(NodeId, CNodeStateStats&)’:
net_processing.cpp:982/1000/1009: error: ‘struct CNodeStateStats’ has no member named ‘nStartingHeight’; did you mean ‘m_starting_height’?
  982 |     stats.nStartingHeight = peer->nStartingHeight;
      |           ^~~~~~~~~~~~~~~
      |           m_starting_height
```
(we'll get there!)

jnewbery commented:
Oops. Sorry Jon. Should be good now.

jnewbery commented:
Rebased

jnewbery commented:
Rebased

jnewbery commented:
This is a slightly difficult rebase since the commits need to be structured a little differently to all build. Will work on it tomorrow.

jnewbery commented:
I'm going to mark this as WIP for now. It'll be a lot easier to rebase after #20624 is merged.

If you want to help with #19398, please review #20624 first!

jnewbery commented:
Rebased. This is now ready for review.

MarcoFalke commented:
```
Running script for: fe391ca326f73337275d726f7de9db384ac35466

sed -i 's/vBlockHashesToAnnounce/m_blocks_for_headers_relay/g' src/net_processing.cpp

sed -i 's/vInventoryBlockToSend/m_blocks_for_inv_relay/g' src/net_processing.cpp

diff --git a/src/net_processing.h b/src/net_processing.h

index 1ad40982b..4998a1e04 100644

--- a/src/net_processing.h

+++ b/src/net_processing.h

@@ -66,9 +66,9 @@ struct Peer {

     /** List of block ids we still have announce.

      * There is no final sorting before sending, as they are always sent

      * immediately and in the order requested. */

-    std::vector<uint256> m_blocks_for_inv_relay GUARDED_BY(m_block_inv_mutex);

+    std::vector<uint256> vInventoryBlockToSend GUARDED_BY(m_block_inv_mutex);

     /** List of headers to relay */

-    std::vector<uint256> m_blocks_for_headers_relay GUARDED_BY(m_block_inv_mutex);

+    std::vector<uint256> vBlockHashesToAnnounce GUARDED_BY(m_block_inv_mutex);

 

     /** This peer's reported block height when we connected */

     std::atomic<int> m_starting_height{-1};

Failed

jnewbery commented:
Thanks @MarcoFalke. Should be fixed now.

Sjors commented on src/net_processing.cpp:
```diff
@@ -1315,13 +1315,17 @@ void PeerManager::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInde
-            pnode->vBlockHashesToAnnounce.push_back(hash);
+    {
+        LOCK(m_peer_mutex);
+        for (auto& it : m_peer_map) {
```
a50dee50fd2aee93b41bfb3b04b24939d0c2f6c1 : `ForEachNode` performs an additional check `NodeFullyConnected` (`pnode->fSuccessfullyConnected && !pnode->fDisconnect;`).

Not sure if that matters much here. Could we prematurely send headers before `VERACK`?

jnewbery replied:
This is an excellent observation. Block hashes could be pushed into `m_blocks_for_headers_relay` before the version/verack handshake is complete. However, we'll never send any blocks inventory before we're fully connected due to the guard clause at the top of SendMessages():

https://github.com/bitcoin/bitcoin/blob/c434e2cca9181ce184efef946ab2ce62f2cd2ee7/src/net_processing.cpp#L4077-L4078

I think this behaviour is ok. The version-verack handshake usually lasts no more than a second or two, and at the worst case times out after a minute, so we're never going to push many block hashes onto this vector.

Sjors commented:
tACK b5b27ea modulo NodeFullyConnected clarification

Sjors reviewed (commented):
tACK b5b27ea modulo NodeFullyConnected clarification

jnewbery commented:
@hebasto @ariard @jonatack - this is now rebased with all comments addressed. Do you have time to rereview?

hebasto commented on src/net_processing.h:
```diff
@@ -61,6 +61,15 @@ struct Peer {
 
+    /** Protects block inventory data members */
+    Mutex m_block_inv_mutex;
+    /** List of block ids we still have announce.
```
a50dee50fd2aee93b41bfb3b04b24939d0c2f6c1

I understand that comment text has been preserved, but maybe
```suggestion
    /** List of block ids we still have to announce.
```

jnewbery replied:
Agree. Will fix.

jnewbery replied:
Fixed

sipa commented:
utACK b5b27ea052905ff7f26a962baa961fcc762c5ee0

Is m_hash_continue protected by any locks? A "this is only accessed by the singular net_processing thread" rationale would be sufficient, but if there happens to be a lock that's held in all access locations, perhaps it's easy to add?

hebasto reviewed (commented):
ACK b5b27ea052905ff7f26a962baa961fcc762c5ee0, tested on Linux Mint 20 (x86_64), I've verified `getpeerinfo` RPC output, and detailed peer info in the "Peers" tab of the "Node window".

sdaftuar commented on src/net_processing.h:
```diff
@@ -72,6 +72,11 @@ struct Peer {
     /** This peer's reported block height when we connected */
     std::atomic<int> m_starting_height{-1};
+    /** The final block hash in an INV message. When the peer requests this
+     * block, we send a BLOCK message to trigger the peer to request the next
```
We send an INV message, not a BLOCK message.

jnewbery replied:
Yes, will fix.

jnewbery replied:
Done.

sdaftuar reviewed (commented):
Concept ACK, just one question about locking and a doc nit.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1315,13 +1315,17 @@ void PeerManager::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInde
+        LOCK(m_peer_mutex);
+        for (auto& it : m_peer_map) {
+            Peer& peer = *it.second;
+            LOCK(peer.m_block_inv_mutex);
```
Would a comment next to `m_peer_mutex` suffice here? This design (take `m_peer_mutex` first, then lock the mutexes protecting the individual data members) is analogous to `ForEachNode()` was doing here before (take `cs_vNodes` first, then lock the mutexes protecting the individual data members in the lambda - here `cs_inventory`).

jnewbery replied:
Added comments to the first commit. Let me know if you think more is required.

jnewbery commented on src/rpc/net.cpp:
```diff
@@ -224,14 +224,14 @@ static RPCHelpMan getpeerinfo()
             }
             obj.pushKV("synced_headers", statestats.nSyncHeight);
             obj.pushKV("synced_blocks", statestats.nCommonHeight);
+            obj.pushKV("startingheight", statestats.nStartingHeight);
```
Will fix.

jnewbery replied:
Fixed

hebasto replied:
Not sure if it is done in the latest push (58f4c8567ba77fbdbd65a12c74628505aacd0d3c).

jnewbery commented:
> Is m_hash_continue protected by any locks? A "this is only accessed by the singular net_processing thread" rationale would be sufficient, but if there happens to be a lock that's held in all access locations, perhaps it's easy to add?

@sipa could we just make it a std::atomic? It doesn't need to be synchronized with any other data, so as long as we're guaranteeing that reads/writes are atomic, I think that's enough?

jnewbery commented:
This has three code review ACKs, with just three minor comments outstanding. I think we can merge now and I can fix those outstanding comments in the next PR in the #19398 sequence. Also happy to fix them here if that's what others prefer.

jnewbery committed:
[net processing] Guard m_continuation_block with m_block_inv_mutex

jnewbery commented:
I've addressed @sdaftuar, @hebasto and @sipa suggestions in the first three commits of #20721.

MarcoFalke commented on src/net_processing.h:
```diff
@@ -66,9 +66,9 @@ struct Peer {
+    std::vector<uint256> m_blocks_for_inv_relay GUARDED_BY(m_block_inv_mutex);
     /** List of headers to relay */
-    std::vector<uint256> vBlockHashesToAnnounce GUARDED_BY(m_block_inv_mutex);
+    std::vector<uint256> m_blocks_for_headers_relay GUARDED_BY(m_block_inv_mutex);
```
22c2b3eff601cbc1aed2e466502bdc5549e11ae2

I don't think this name is correct. Headers relay isn't implied. It might very well fall back to inv relay. Most importantly when the peer asked us to. Though, there are other reasons, too.

jnewbery replied:
We hope to announce these blocks via headers. If we can't, we'll move the hashes onto the `m_blocks_for_inv_relay` and then announce via inv.

I've updated the comment. Let me know if you think it's clear enough now.

MarcoFalke replied:
Thanks. Looks good now.

MarcoFalke commented on src/net_processing.h:
```diff
@@ -72,6 +72,11 @@ struct Peer {
 
     /** This peer's reported block height when we connected */
     std::atomic<int> m_starting_height{-1};
+    /** The final block hash in an INV message. When the peer requests this
```
b5b27ea052905ff7f26a962baa961fcc762c5ee0

I don't think this is true either. `m_hash_continue` is a purely internal helper, never sent over the wire directly.  Instead, the tip hash is sent to trigger the peer to ask further.

jnewbery replied:
I've updated the comment to be "The final block hash that we sent in an `inv` message to this peer." Let me know if you have suggestions for better wording.

MarcoFalke replied:
Thanks. Looks good now.

maflcko reviewed (commented):
review ACK b5b27ea052905ff7f26a962baa961fcc762c5ee0 I reviewed the code and it probably wont crash, but I had a few questions 🚟

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK b5b27ea052905ff7f26a962baa961fcc762c5ee0 I reviewed the code and it probably wont crash, but I had a few questions 🚟
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUioOwv6Amw0DZDkCn50bwGUmV7m2mPj2cL5zV9XhpbPUhiMa/emED2JTAu020/z
LQZQp0pzHsSOCtBzVbfbi4GTK45AEfAV2FEMgMaQNO0QQIdFXD86angiyLZuojbM
uq5gnQS9d2uPHgfbWpnGVY7VEMIrIt5wDwXi81geVErEYT3yIf6d7sKazeq0Rw2t
VIIYUsWzESgJpw7iSMVuqP/Du9zuIkzreaLAi3VyJ6BqVxPDKF0OTVE4rsPh0UTN
UAYyz0TczihypdAUm1KQGYeLuHORg+OpDf0ErRwdrahzz1Pk6i6qceJt6Xiaatmy
F0VvaGCxRHCfZp2OGjAmp+Dv8v5AWLYnx4h4dobg+IleEVfOnQEp7G6guB/z30Ht
6mRu6x3PkFJH/LSfSmxRQLuJoteM+ZaPpm15hPrT+eismtk77EslGYDjo4SiUNxM
4u+r8fycyixslocpoByX8wTF0lrsY+YYH5hQINL1UIuQY70qPq9GLQlsTuJAZQyk
nNGzdlSe
=l+a9
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `aea33e795fcbc5727e94f7d8ccebdce44be9f051f6b87be26df0c7f774cd7c32  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108aea33e795fcbc5727e94f7d8ccebdce44be9f051f6b87be26df0c7f774cd7c32f0101bcf7cf17a29a9f5a03a240c45ea4edc08fff0106a6d4f3242fc34f2af09a265b584bfd608f1045fddfd0cf008758c0121a54017b90083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0108b111ba40a248085b39def59c82795f208f1045fddfd0bf008d6537e1dde636dee0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010c2d6d5b02a3a04d3973580a17235cd6f08f020d5e09922b0bbd8073bc63499080f1381cebae7832388380bcfcf05e4da63a15308f1045fddfd0cf008e5ad6c563fe78daa0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f81c5f08da7acee3b67aceb9729a3aab08f02007c746de92afcb55bdd0921f695970d454b8ff0350030f926e028bdaa285e99508f1045fddfd0cf008f0d74659286ae42c0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

jnewbery commented:
Thanks for the review @MarcoFalke . There are enough review comments here that it makes sense to address them in this PR instead of a follow-up. It also looks to me like this isn't entirely a clean merge, so I'm going to rebase on master.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1843,7 +1844,8 @@ void PeerManager::ProcessHeadersMessage(CNode& pfrom, const std::vector<CBlockHe
             // TODO: optimize: if pindexLast is an ancestor of ::ChainActive().Tip or pindexBestHeader, continue
             // from there instead.
-            LogPrint(BCLog::NET, "more getheaders (%d) to end to peer=%d (startheight:%d)\n", pindexLast->nHeight, pfrom.GetId(), pfrom.nStartingHeight);
+            PeerRef peer = GetPeerRef(pfrom.GetId());
```
Yes, passing in from outside is a much better idea. I've done that now.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1315,13 +1315,17 @@ void PeerManager::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInde
-            pnode->vBlockHashesToAnnounce.push_back(hash);
+    {
+        LOCK(m_peer_mutex);
+        for (auto& it : m_peer_map) {
```
I originally had one but removed it because it didn't seem worth it. We can revisit that decision later if it seems useful.

ForEachNode was useful because it allowed passing a lambda into CConnman which would be executed while the cs_vNodes lock was held. Here, the m_peer_mutex is in PeerMan, so it can be held directly.

jnewbery commented on src/net_processing.h:
```diff
@@ -61,6 +61,15 @@ struct Peer {
+     * There is no final sorting before sending, as they are always sent
+     * immediately and in the order requested. */
+    std::vector<uint256> vInventoryBlockToSend GUARDED_BY(m_block_inv_mutex);
+    /** List of headers to relay */
```
Ok, I've added the word unfiltered back.

jnewbery commented:
Addressed all comments.

> Is m_hash_continue protected by any locks? A "this is only accessed by the singular net_processing thread" rationale would be sufficient, but if there happens to be a lock that's held in all access locations, perhaps it's easy to add?

@sipa I've added a commit to guard m_hash_continue with m_block_inv_mutex.

jonatack commented on src/rpc/net.cpp:
```diff
@@ -224,14 +224,14 @@ static RPCHelpMan getpeerinfo()
             }
             obj.pushKV("synced_headers", statestats.nSyncHeight);
             obj.pushKV("synced_blocks", statestats.nCommonHeight);
+            obj.pushKV("startingheight", statestats.m_starting_height);
```
In 2f023a8c0e445985a20, `startingheight` should remain before `synced_headers` and `synced_blocks` (here and in the help) for consistency with the GUI peers info (or maybe less preferably, the GUI should be updated to follow this change)

jnewbery replied:
done

jonatack commented on src/net_processing.h:
```diff
@@ -60,6 +63,25 @@ struct Peer {
+     * When the peer requests this block, we send an `inv` message to trigger
+     * the peer to request the next sequence of block hashes.
+     * Most peers use headers-first syncing, which doesn't use this mechanism */
+    uint256 m_hash_continue GUARDED_BY(m_block_inv_mutex) {};
```
25ccfd3b while renaming `hashContinue`, I wonder if it can be better named (but don't have one to propose)

jnewbery replied:
Changed to `m_continuation_block`

jonatack reviewed (commented):
ACK 9aad3e4f3347990641da23c0d9a3f05e1669cabf modulo the RPC fields/help order (see comment below)

practicalswift commented:
Concept ACK

MarcoFalke commented on src/net_processing.h:
```diff
@@ -46,6 +46,8 @@ struct CNodeStateStats {
  * Memory is owned by shared pointers and this object is destructed when
  * the refcount drops to zero.
  *
+ * Mutexes inside this struct must not be held when locking m_peer_mutex.
```
717a374e74b64b7b90bc1b2995e8900212bd0bfe

nit to avoid the inversion and give a rationale:

"m_peer_mutex should be locked before locking a mutex inside this struct to avoid lock order violations"

This will also weaken the requirement, because for some mutexes it could make sense to not require m_peer_mutex beforehand.

jnewbery replied:
That isn't precise enough. It's fine to lock m_peer_mutex, take a shared_ptr to one of the Peer objects, release m_peer_mutex and later take one of the inner locks. The only rule is that we don't want to take m_peer_mutex while already holding one of the inner locks.

This is the same as cs_vNodes and the inner locks in CNode.

MarcoFalke replied:
Ah right. I didn't parse this correctly. Thanks for the explanation.

MarcoFalke commented on src/net_processing.h:
```diff
@@ -210,7 +212,8 @@ class PeerManager final : public CValidationInterface, public NetEventsInterface
 
-    /** Protects m_peer_map */
+    /** Protects m_peer_map. This mutex must not be locked while holding a lock
+     *  on any of the mutexes inside a Peer object. */
```
same

jnewbery replied:
as above

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -1615,15 +1615,17 @@ void static ProcessGetBlockData(CNode& pfrom, Peer& peer, const CChainParams& ch
-            vInv.push_back(CInv(MSG_BLOCK, ::ChainActive().Tip()->GetBlockHash()));
-            connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::INV, vInv));
-            peer.m_hash_continue.SetNull();
+        LOCK(peer.m_block_inv_mutex); {
```
What syntax is this? I don't think we use `LOCK(); { ... }` anywhere in the codebase

MarcoFalke replied:
Also, doesn't this lock violate the documentation you added in the first commit?

jnewbery replied:
This is my brain not working properly. I've fixed the syntax.

This doesn't violate the documentation. That says that it's not permitted to lock m_peer_map while holding an inner mutex. This is locking an inner mutex.

maflcko reviewed (commented):
review ACK 9aad3e4f33 🔱

Changes since previous review:

* Add lock-order docs
* Pass in Peer& to ProcessHeadersMessage
* Shuffle order in RPC doc
* Rebase
* Doc fixups
* New commit with a lock


<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 9aad3e4f33 🔱

Changes since previous review:

* Add lock-order docs
* Pass in Peer& to ProcessHeadersMessage
* Shuffle order in RPC doc
* Rebase
* Doc fixups
* New commit with a lock

-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhhwgv+M611eBCdMzaHcsl/tL05mOFVC94Nc9F7t5MCW8iuNlEWaT7yzpQaPV5g
8L9MZqFGi7YpPVrL2y12ky0MYBfG9MdploG2F53O4WvLnyEa6PXw04HZKqlv9yfS
qcJEwjJc+XvKPohl6NZg91GDyofVttKI8bqCqjPhEbZYCi2+RQ9o0OZ1PJEr8b6L
KgtXpj+YgtpRwpH5ny/E/mR5LwfLFmxYNmgQlEpMvjIqyxcSUgrYXn77vxMJI//T
Y2g5NowkMo43S12fT6KWJk9eMoRy+L2kgrCD9LI0r/wuY1B+9rCamBgzeBKIuk6f
EfFnS33FIG8xcT21+x5g38J6jlHz6HlSy/huA31ALBhDItMTR+dm7VI+1KXKTTag
nhN9S/ZVWXjjPaZKzbAmKguPRbJWdOcoWdyAlZy+S4qeC+wdEqT+GBGvhj0TZ3du
56VMY6sNhdG8kUQr4KuwugereUtKQDJvHMehBH6ki6br/G6W71kdpCYPhxK5fZkx
khylqj75
=v6fJ
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `431bceaaa7cf87a17743f11d77f4a61c5cdf9d1374a5f4fcd4e5683b564af330  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108431bceaaa7cf87a17743f11d77f4a61c5cdf9d1374a5f4fcd4e5683b564af330f0107915b6b8d827d79936bc1131f412a07908fff01091fbb47ef010d88b9a83319fbe65c27e08f1045fdf0917f0080426a93e18cc52540083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff010b136666c2895836c78ef1d7c183c9e5a08f1045fdf0917f008a02d029412ccaa570083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010ce9b323d44365152637910f2d4e6b41108f1045fdf0917f008eba83d9f76e3e7ae0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010f4ede23f13d23066d3076256fc380eba08f1045fdf0918f008565d4687765285d40083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1764,7 +1775,9 @@ void PeerManager::SendBlockTransactions(CNode& pfrom, const CBlock& block, const
-void PeerManager::ProcessHeadersMessage(CNode& pfrom, const std::vector<CBlockHeader>& headers, bool via_compact_block)
+void PeerManager::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
+                                        const std::vector<CBlockHeader>& headers,
+                                        bool via_compact_block)
```
This is permitted by our clang formatter. I find it easier to read lines that are less than 100 columns, so I think this is a matter of taste.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -1854,7 +1867,8 @@ void PeerManager::ProcessHeadersMessage(CNode& pfrom, const std::vector<CBlockHe
             // from there instead.
-            LogPrint(BCLog::NET, "more getheaders (%d) to end to peer=%d (startheight:%d)\n", pindexLast->nHeight, pfrom.GetId(), pfrom.nStartingHeight);
+            LogPrint(BCLog::NET, "more getheaders (%d) to end to peer=%d (startheight:%d)\n",
+                                 pindexLast->nHeight, pfrom.GetId(), peer.m_starting_height);
```
Again, this is personal taste, so I'm going to resolve this.

jnewbery commented on src/net_processing.h:
```diff
@@ -180,7 +202,9 @@ class PeerManager final : public CValidationInterface, public NetEventsInterface
     void ProcessOrphanTx(std::set<uint256>& orphan_work_set) EXCLUSIVE_LOCKS_REQUIRED(cs_main, g_cs_orphans);
     /** Process a single headers message from a peer. */
-    void ProcessHeadersMessage(CNode& pfrom, const std::vector<CBlockHeader>& headers, bool via_compact_block);
+    void ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
```
as above

jnewbery commented:
Thanks for the latest reviews @jonatack and @MarcoFalke. I've addressed/responded to all of your comments.

I'd like to stabilize this PR now unless any functional issues are found. We have ACKs on the code change at various times from @MarcoFalke, @jonatack, @hebasto, @sipa and @Sjors. If there are any further style suggestions, I can address them in the next PR in this series.

maflcko reviewed (approved):
re-ACK 3002b4af2b4fde63026f8f7c575452a5c989c662 🌓

only changes:

* Move RPC help once more
* Fix LOCK scope {} nit
* Rename a member


<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 3002b4af2b4fde63026f8f7c575452a5c989c662 🌓

only changes:

* Move RPC help once more
* Fix LOCK scope {} nit
* Rename a member

-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUirewv/WiQMgzv/dZCAqCSBX8qa6mPZiEw4j5mOPAnylVH3P+qiQGEXrHpKVjTj
XO0rGPI+dEvej90YVSThX231sBS1VXvP4hv8JKYHVdoAVP2j3XbmEmtZrPoGqViR
eMtblmUALv5A8OfkquBDUb/PVDr9FxUycTe01tGxyxhcLtaTfoIcq+6Elh7NkWMm
tgszY1rz2S7pjYU3GH/an+v5BYOBAi3HG5vHETEICQGjabAEduMZFF2bueQZIj1Q
CkVhnj/mcFx7Fypc0J1q5LsPonuPHVfqMK38dO6iFxcpLfLZKwmcnXAJ7OFnHkQF
JJ1zXA5HRixl5UKakZqUkjhjroD71vfODzfhSLaszEZLqeGXx0+N17CAd8Vgblk2
ReLskQfzmco+aHWX6Hl7CKj7+yTszpWTMvmMvkl08EaNnAGuGBbF1TgI8oA213Ry
QwjwlP5Ohx7Iy33qtysxAq5/4jqypu/srlhxcfFCidi+AvujeZqcyXLztxw54zLj
cWdIBtpT
=PVp9
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `03268669e27bdf510ed13b406639896a3bedb0573f1b2d13c4c88ed636966237  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010803268669e27bdf510ed13b406639896a3bedb0573f1b2d13c4c88ed636966237f010ecbee7207f542127ae3757cd4980cb8b08fff01054ccb28600eba8f30afd62eced73aa0e08f1045fdf3efbf008390cb2820dc0f17e0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010811d165e3c331a6cad36f2e5e114693a08f1045fdf3efcf00897b3add0db76de2e0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010b1d32d4a463bd63af72357bb1def4cca08f1206ddb15f798547b319ad0398ebec78b29acc00638b804e67ac9fa9e791a2e16cf08f1045fdf3efbf008c821995727acae9f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6df010e0e8d6b2df797b4e3c48ffb72765ec4008f1045fdf3efbf008a4048f2b0220c21d0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

jonatack commented:
re-ACK 3002b4af2b4fde63026f8f7c575452a5c989c662 per `git diff 9aad3e4 3002b4a`

Sjors commented:
Code review re-ACK 3002b4af2b4fde63026f8f7c575452a5c989c662

kaycoinsUSA reviewed (commented):
Done
